iconfetcher.go

```diff
@@ -0,0 +1,130 @@
+package jujusvg
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+
+	"github.com/juju/utils/parallel"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charm.v5"
+)
+
+// An IconFetcher provides functionality for retrieving icons for the charms
+// within a given bundle.  The FetchIcons function accepts a bundle, and
+// returns a map from charm paths to icon data.
+type IconFetcher interface {
+	FetchIcons(*charm.BundleData) (map[string][]byte, error)
+}
+
+// LinkFetcher fetches icons as links so that they are included within the SVG
+// as remote resources.
+type LinkFetcher struct {
+	// IconURL returns the URL of the entity for embedding
+	IconURL func(*charm.Reference) string
+}
+
+// FetchIcons generates the svg image tags given an appropriate URL, generating
+// tags only for unique icons.
+func (l *LinkFetcher) FetchIcons(b *charm.BundleData) (map[string][]byte, error) {
+	// Maintain a list of icons that have already been fetched.
+	alreadyFetched := make(map[string]bool)
+
+	// Build the map of icons.
+	icons := make(map[string][]byte)
+	for _, serviceData := range b.Services {
+		charmId, err := charm.ParseReference(serviceData.Charm)
+		if err != nil {
+			return nil, errgo.Notef(err, "cannot parse charm %q", serviceData.Charm)
+		}
+		path := charmId.Path()
+
+		// Don't duplicate icons in the map.
+		if !alreadyFetched[path] {
+			alreadyFetched[path] = true
+			icons[path] = []byte(fmt.Sprintf(`
+				<svg xmlns:xlink="http://www.w3.org/1999/xlink">
+					<image width="96" height="96" xlink:href=%q />
```

Comment:
How about using "%s" and xml.EscapeText instead of %q?
You'll probably want to wrap xml.EscapeText as it's
awkwardly defined. Something like this.

    func escapeString(s string) string {
         var buf bytes.Buffer
         xml.EscapeText(&buf, []byte(s))
         return buf.String()
     }

Then you don't have to worry about the possible things that IconURL
might return.